terraform/main.tf

```diff
@@ -0,0 +1,186 @@
+variable "aws_region" { default = "us-west-2" }
+variable "aws_profile" { default = "default" }
+variable "aws_vpc" { default = "vpc-41d45124" }
+variable "aws_subnet" { default = "subnet-7424b611" }
+variable "chef_server_instance_type" { default = "m5.xlarge" }
+variable "automate_server_instance_type" { default = "m5.2xlarge" }
+variable "default_security_group" { default = "sg-c9beb2ac" }
+variable "aws_ami_user" { default = "centos" }
+variable "aws_ami_id" { default = "" }  # leave blank to auto-select the latest highperf CentOS 7 image
+variable "aws_instance_names" { default = ["automate-server", "chef-server"] }
+variable "aws_instance_types" { default = ["m4.xlarge", "m4.2xlarge"] }
+variable "aws_key_pair_name" { }
+variable "aws_key_pair_file" { }
+variable "tag_dept" { }
+variable "tag_contact" { }
+variable "container_username" { }
+variable "container_uid" { }
+variable "container_gid" { }
+variable "container_data_mount" { }
+variable "automate_enterprise" { }
+variable "automate_admin_password" { }
+variable "docker_host_ip" { }
+variable "automate_enabled" { }
+variable "automate_token" { }
+variable "chef_server_docker_origin" { }
+variable "automate_docker_origin" { }
+variable "chef_server_version" { }
+variable "automate_version" { }
+variable "chef_server_org" { }
+
+provider "aws" {
+  region  = "${var.aws_region}"
+  profile = "${var.aws_profile}" // uses ~/.aws/credentials by default
+}
+
+resource "random_id" "cluster_id" { byte_length = 4 }
+
+data "aws_ami" "centos" {
+  most_recent = true
+  owners = ["446539779517"]
+
+  filter {
+    name   = "name"
+    values = ["chef-highperf-centos7-*"]
+  }
+}
+
+data "template_file" "env_sh" {
+  template = "${file("${path.module}/env.tpl")}"
+
+  vars {
+    container_username = "${var.container_username}"
+    container_uid = "${var.container_uid}"
+    container_gid = "${var.container_gid}"
+    container_data_mount = "${var.container_data_mount}"
+    automate_enterprise = "${var.automate_enterprise}"
+    automate_admin_password = "${var.automate_admin_password}"
+    automate_enabled = "${var.automate_enabled}"
+    automate_token = "${var.automate_token}"
+    chef_server_docker_origin = "${var.chef_server_docker_origin}"
+    chef_server_version = "${var.chef_server_version}"
+    automate_docker_origin = "${var.automate_docker_origin}"
+    automate_version = "${var.automate_version}"
+  }
+}
+
+data "template_file" "group" {
+  template = "${file("${path.module}/group.tpl")}"
+
+  vars {
+    container_username = "${var.container_username}"
+    container_gid = "${var.container_gid}"
+  }
+}
+
+data "template_file" "passwd" {
+  template = "${file("${path.module}/passwd.tpl")}"
+
+  vars {
+    container_username = "${var.container_username}"
+    container_uid = "${var.container_uid}"
+    container_gid = "${var.container_gid}"
+  }
+}
+
+resource "aws_instance" "automate_cluster" {
+  count = 2
+  # count.index
+  # 0 Automate Server
+  # 1 Chef Server
```

Comment:
I get why you'd want to do this and it is super clever, but TBH I don't love it because it is *too* clever.    I wouldn't block merging on it, but I worry that future travelers who happen upon this code will be confused by it.

Reply:
The biggest wins:
 * it cuts the total provisioning time (docker pulls) in half since it runs in parallel for each server, not sequentially
 * it reduces redundant provision sections

I can take it out though if those are outweighed by readability.